### ISSUE TITLE
Trying to use a FileOpenPicker while running the app as Administrator will crash the app

### ISSUE THREAD

**Issue by bogdan-patraucean** [microsoft/WindowsAppSDK#2504]
### Describe the bug

I am trying to open a **File Picker** but it works only if I am not using the app in elevation mode.

I'm getting the following error: 

`System.Runtime.InteropServices.COMException: 'Error HRESULT E_FAIL has been returned from a call to a COM component.'`

The error appears when the code tries to execute the following line of code:

`var file = await picker.PickSingleFileAsync();`

### Steps to reproduce the bug

1. Create a WASDK app using version **1.4.2** and **.NET 7**
2. Create an extension method so you can get the Window Handle

_AppWindowExtension.cs_
```
public static IntPtr GetHandle(this Microsoft.UI.Xaml.Window window)
{
    return WinRT.Interop.WindowNative.GetWindowHandle(window);
}
```

3. Make the MainWindow available outside the App class

_App.xaml.cs_

```
public static Window MainWindow { get; private set; }
```

4. Use the generated **Button_Click** event method to open the **File Picker**

_MainWindow.xaml.cs_

```
var picker = new FileOpenPicker();
picker.FileTypeFilter.Add("*");

InitializeWithWindow.Initialize(picker, App.MainWindow.GetHandle());

var file = await picker.PickSingleFileAsync();
```

5. Build the app and try to **Run as Administrator** (it's not neccesary to configure elevation, just right click on the app works)
6. **Click** the button to open the File Picker
7. The app will crash :(

### Expected behavior

I would expect this basic feature of picking a file to work even if I run the app as admin. 
My app has a lot of features that require elevation so getting rid of elevation is not an option for me. 

### NuGet package version

1.4.2

### Packaging type

Packaged (MSIX)

### Windows version

Insider Build (xxxxx) (only on Windows 11)

### IDE

Visual Studio 2022

### Additional context

The same for FolderPicker

**Comment by Petrarca181** [comment]
I can confirm this bug. Any temporary workground for this?

**Comment by bogdan-patraucean** [comment]
@Petrarca181 yes, you can use directly PInvoke with CsWin32, but it's too much overhead if you ask me.

**Comment by lukedukeus** [comment]
@Petrarca181 I had to write my own with P/Invoke as well, It's basic and blocking / non-async, but gets the job done. Mostly copied from [here](https://stackoverflow.com/a/68712025/10197738).


```
using System;
using System.Runtime.InteropServices;

namespace Namespace
{

    [StructLayout(LayoutKind.Sequential, CharSet = CharSet.Auto)]
    public struct OpenFileName
    {
        public int lStructSize;
        public IntPtr hwndOwner;
        public IntPtr hInstance;
        public string lpstrFilter;
        public string lpstrCustomFilter;
        public int nMaxCustFilter;
        public int nFilterIndex;
        public string lpstrFile;
        public int nMaxFile;
        public string lpstrFileTitle;
        public int nMaxFileTitle;
        public string lpstrInitialDir;
        public string lpstrTitle;
        public int Flags;
        public short nFileOffset;
        public short nFileExtension;
        public string lpstrDefExt;
        public IntPtr lCustData;
        public IntPtr lpfnHook;
        public string lpTemplateName;
        public IntPtr pvReserved;
        public int dwReserved;
        public int flagsEx;
    }
    public static class FilePicker
    {

        [DllImport("comdlg32.dll", SetLastError = true, CharSet = CharSet.Auto)]
        private static extern bool GetOpenFileName(ref OpenFileName ofn);

        // usage: string filename = FilePicker.ShowDialog("C:\\", new string[] { "png", "jpeg", "jpg" }, "Image Files", "Select an Image File...");
        public static string ShowDialog(string startingDirectory, string[] filters, string filterName, string dialogTitle)
        {
            var ofn = new OpenFileName();
            ofn.lStructSize = Marshal.SizeOf(ofn);

            ofn.lpstrFilter = filterName;
            foreach (string filter in filters)
            {
                ofn.lpstrFilter += $"\0*.{filter}";
            }
            ofn.lpstrFile = new string(new char[256]);
            ofn.nMaxFile = ofn.lpstrFile.Length;
            ofn.lpstrFileTitle = new string(new char[64]);
            ofn.nMaxFileTitle = ofn.lpstrFileTitle.Length;
            ofn.lpstrTitle = dialogTitle;
            if (GetOpenFileName(ref ofn))
                return ofn.lpstrFile;
            return string.Empty;
        }
    }
}

```

Then, you can call it like so:

`string filename = FilePicker.ShowDialog("C:\\", new string[] { "png", "jpeg", "jpg" }, "Image Files", "Select an Image File...");`


Or, if you must use naitive, you can work around this bug by running the rest of your code as admin, and this not as admin using impersonation.

**Comment by AdamBraden** [comment]
@manodasanW - crashes right at the PickSingleFileAsync() call.

**Comment by derekhe** [comment]
Yes, the problem still exist in WinUI3 with admin right. Please fix this!

**Comment by AdamBraden** [comment]
After discussing this with the filesystem team, the recommendation is to use the IFileOpenDialog apis from CSWin32. Here is a recent thread discussing this very topic - [FileOpenDialog in CsWin32](https://github.com/microsoft/CsWin32/discussions/671).

**Comment by bogdan-patraucean** [comment]
@AdamBraden thank you for oferring this workaround. Any updates related to fixing the issue?

**Comment by AdamBraden** [comment]
What we could consider is creating a WinAppSDK wrapper around IFileDialog apis that provide an equivalent to the Picker apis. There are some opportunities to improve as well - StorageFiles are very heavyweight objects, and so we could consider returning something more optimized.

**Comment by bogdan-patraucean** [comment]
@manodasanW, @MikeHillberg, @AdamBraden, hi! Any progress on this? I would appreciate an update. I'm curious how it goes.

**Comment by derekhe** [comment]
WinUI3 makes me crazy for even basic stuff is not working. I return back to use electron.

**Comment by bogdan-patraucean** [comment]
Indeed, such a simple thing like a FilePicker requires advanced knowledge and years of experience to understand. I'm a C# developer and I need to use unmanaged code with C++ to open a file picker...

**Comment by bogdan-patraucean** [comment]
I just discovered that on Windows 10 it works. The problem is only on Windows 11.

**Comment by fredemmott** [comment]
Same problem on windows 10 here.

**Comment by bogdan-patraucean** [comment]
> Same problem on windows 10 here.

what version do you have?

**Comment by ivberg** [comment]
same issue here on Win11

**Comment by bogdan-patraucean** [comment]
It's been 9 months since the issue was reported, any hopes it will be fixed in WASDK 1.3?

**Comment by bogdan-patraucean** [comment]
@fredemmott tested again on a clean installation of Windows 10 and no longer works

**Comment by AdamBraden** [comment]
The workaround is to use IFileDialog. For example:

            try 
            {
                PInvoke.CoCreateInstance(
                    typeof(FileOpenDialog).GUID,
                    null,
                    CLSCTX.CLSCTX_INPROC_SERVER,
                    out IFileDialog fileOpenDialog).ThrowOnFailure();
                fileOpenDialog.Show((HWND)App.MainWindow.GetHandle());

                fileOpenDialog.GetResult(out IShellItem item);
                item.GetDisplayName(SIGDN.SIGDN_FILESYSPATH, out PWSTR name);
                Console.WriteLine(name.ToString());
                Marshal.FreeCoTaskMem(new IntPtr(name.Value));
            }
            catch (COMException)
            {
                // Handle canceled.
            }

**Comment by bogdan-patraucean** [comment]
@AdamBraden what if you want to filter the file types? I tried to use IFileDialog but I did not manage to use filters.

**Comment by AdamBraden** [comment]
You tried SetFilter? https://learn.microsoft.com/en-us/windows/win32/api/shobjidl_core/nf-shobjidl_core-ifiledialog-setfilter

**Comment by bogdan-patraucean** [comment]
@AdamBraden yes, thank you for answering, but I couldn't understand what values to pass as parameters. If you could give me an example, I would really apreciate it.

**Comment by fredemmott** [comment]
https://learn.microsoft.com/en-us/windows/win32/api/shobjidl_core/nf-shobjidl_core-ifiledialog-setfiletypes might be what you want instead

https://github.com/OpenKneeboard/OpenKneeboard/blob/e45bbfd46d953a562a9aaa59f38607472d9ea830/src/app/app-winui3/FilePicker.cpp#L50 is a C++ example; I don't have a C# one.

In short, "name" in the struct is user-visible name (e.g. "Text File"), "pattern" is semicolon-separated globs, e.g. `*.txt;*.log`

**Comment by bogdan-patraucean** [comment]
@fredemmott yes, that'a exactly how I tried I just wasn't able to find a fix for all the type errors I got.

**Comment by castorix** [comment]
> You tried SetFilter? https://learn.microsoft.com/en-us/windows/win32/api/shobjidl_core/nf-shobjidl_core-ifiledialog-setfilter

Replaced by SetFileTypes
(I had posted samples for IFileDialog on MS forums, in C# and VB.NET)

**Comment by bogdan-patraucean** [comment]
@castorix link please?

**Comment by castorix** [comment]
> @castorix link please?

An old one in Winforms  [C# transfer file from Android device to pc using c#](https://learn.microsoft.com/en-us/answers/questions/329383/c-transfer-file-from-android-device-to-pc-using-c?childtoview=329388&orderby=oldest)
It is same code in WinUI 3 for P/Invoke.  It has to be improved...

**Comment by castorix** [comment]
> @castorix link please?

I found old links... but they cutted the code :-(
I will try to re-post, but you can find various implementations by typing for example in Google "IFileOpenDialog  comimport github"

**Comment by Jay-o-Way** [comment]
@AdamBraden I noticed even WinUI 3 Gallery has the same problem in their FilePicker example - up til now. Running the app as Admin makes the function break. Conclusion: it's a WinUI/SDK bug. When will the actual problem be solved? :)

**Comment by Petrarca181** [comment]
> @AdamBraden I noticed even WinUI 3 Gallery has the same problem in their FilePicker example - up til now. Running the app as Admin makes the function break. Conclusion: it's a WinUI/SDK bug. When will the actual problem be solved? :)

Probably in 2-3 years.

**Comment by AdamBraden** [comment]
As mentioned above the Picker apis are not designed for Admin. 

Because the storage apis return heavy weight objects, a better approach would be to use IFileDialog. A good example has been published here:
https://learn.microsoft.com/en-us/uwp/api/windows.storage.pickers.filesavepicker?view=winrt-22621#in-a-desktop-app-that-requires-elevation

**Comment by fredemmott** [comment]
Is this a wontfix then? If so, at a minimum, that documentation should also be on the fileopenpicker documentation

**Comment by Levminer** [comment]
> As mentioned above the Picker apis are not designed for Admin.
> 
> Because the storage apis return heavy weight objects, a better approach would be to use IFileDialog. A good example has been published here: [learn.microsoft.com/en-us/uwp/api/windows.storage.pickers.filesavepicker?view=winrt-22621#in-a-desktop-app-that-requires-elevation](https://learn.microsoft.com/en-us/uwp/api/windows.storage.pickers.filesavepicker?view=winrt-22621#in-a-desktop-app-that-requires-elevation)

In my opinion, if I have to write 150 lines just to get a dialog in 2023, I just never use this SDK again.

**Comment by Petrarca181** [comment]
Really? This is modern SDK or what?  I don't care if it's returning heavy objects or other objects. Run windows programs as admin is a normal thing many apps need it .

**Comment by yaira2** [comment]
> As mentioned above the Picker apis are not designed for Admin.

It seems that this is causing a lot of friction with devs, if possible, perhaps the apis need to be rethought to support this scenario.

**Comment by bogdan-patraucean** [comment]
```
public unsafe string PickFile(string filter)
    {
        try
        {
            string path = null;

            PInvoke.CoCreateInstance(typeof(FileOpenDialog).GUID, null, CLSCTX.CLSCTX_INPROC_SERVER,out IFileOpenDialog picker).ThrowOnFailure();

            List<COMDLG_FILTERSPEC> extensions = new();

            if (!string.IsNullOrEmpty(filter))
            {
                string[] tokens = filter.Split('|');

                if (0 == tokens.Length % 2)
                {
                    for (int i = 1; i < tokens.Length; i += 2)
                    {
                        COMDLG_FILTERSPEC extension;

                        extension.pszSpec = (char*)Marshal.StringToHGlobalUni(tokens[i]);
                        extension.pszName = (char*)Marshal.StringToHGlobalUni(tokens[i - 1]);
                        extensions.Add(extension);
                    }
                }
            }

            picker.SetFileTypes(extensions.ToArray());
            picker.Show((HWND)App.Window.Hwnd());
            picker.GetResult(out IShellItem item);
            item.GetDisplayName(SIGDN.SIGDN_FILESYSPATH, out PWSTR name);

            path = name.ToString();
            Marshal.FreeCoTaskMem(new IntPtr(name.Value));

            return path;
        }
        catch (Exception)
        {
            return null;
        }
    }
```

I managed to do a method that kinda work, so until the issue is fixed, this ugly workaround that requires CsWin32 with PInvoke will do the job. It's not ideal, but an ugly solution it's better than a broken feature.

**Comment by Petrarca181** [comment]
> // filter value guide: "[name]|[extension]|[name2]|[extension2]"
> // name can be whatever you want but not null
> public unsafe string PickFile(string filter)
>     {
>         try
>         {
>             string path = null;
> 
>             PInvoke.CoCreateInstance(typeof(FileOpenDialog).GUID, null, CLSCTX.CLSCTX_INPROC_SERVER,out IFileOpenDialog picker).ThrowOnFailure();
> 
>             List<COMDLG_FILTERSPEC> extensions = new();

Can you post usings?  I'm  not so much experience and can't in any way to find CoCreateInstance   in PInvoke what reference I should add?

**Comment by castorix** [comment]
> @bogdan-patraucean Can you post usings? I'm not so much experience and can't in any way to find CoCreateInstance in PInvoke what reference I should add?

I just uploaded a small test sample (with a C# class from C++ headers that I had done in WinForms) : https://github.com/castorix/WinUI3_IFileDialog

**Comment by charifield** [comment]
I was encountering a similar problem when trying to use the FolderPicker when running as admin. Nothing would appear. So for anyone that was searching for a solution with a problem with FolderPicker, here is how I got a folder picker working when running as admin:

```
[DllImport("Shell32.dll", CharSet = CharSet.Unicode)]
      public static extern IntPtr SHBrowseForFolder(ref BROWSEINFO lpbi);

      [DllImport("Shell32.dll", CharSet = CharSet.Unicode)]
      public static extern bool SHGetPathFromIDList(IntPtr pidl, IntPtr pszPath);

      [StructLayout(LayoutKind.Sequential, CharSet = CharSet.Unicode)]
      public struct BROWSEINFO
      {
          public IntPtr Owner;
          public IntPtr PIDL;
          public IntPtr DisplayName;
          public string Title;
          public uint Flags;
          public IntPtr Callback;
          public IntPtr Param;
          public int Image;
      }

      public const uint BIF_NEWDIALOGSTYLE = 0x00000040;
      public const uint BIF_USENEWUI = BIF_NEWDIALOGSTYLE | 0x00000200;

      private void btnBrowse_Click(object sender, RoutedEventArgs e)
      {
          var bi = new BROWSEINFO
          {
              Title = "Select Folder",
              Flags = BIF_NEWDIALOGSTYLE | BIF_USENEWUI,
          };

          var pidl = SHBrowseForFolder(ref bi);
          if (pidl != IntPtr.Zero)
          {
              var path = Marshal.AllocHGlobal(260);
              SHGetPathFromIDList(pidl, path);
              Console.WriteLine("Selected folder: " + Marshal.PtrToStringUni(path));
              Marshal.FreeHGlobal(path);
          }
          else
          {
              Console.WriteLine("Operation cancelled.");
          }
      }
```

**Comment by DarranRowe** [comment]
Just to mention it here. This error is fully reproducible in a desktop Windows API application without the Windows App SDK being loaded or even referenced.
This, along with the fact that you cannot use the pickers from desktop applications (which WinUI 3 is) that are packaged up as partial trust means that the UWP pickers are very problematic to use and really need to be fixed in Windows.

**Comment by anilmawji** [comment]
Just ran into this issue. In my case the application does not crash or close, instead the file picker has no effect.

**Comment by Saiteja-Rangi** [comment]
I faced a similar problem which you can read about in this link [here](https://stackoverflow.com/q/76858694/18223845). After looking into it, I figured out why it was happening and found a way to fix it. It seems the issue is due to how permissions are handled.

Here's what I found: When I first installed the app without using "Run as Administrator," it set up the app data folder with permissions tied to the user account I was logged in with. Later, when I tried running the app as an administrator, it crashed when I tried to use the file picker.

The problem is that the app treats administrator privileges as if they belong to a different user, not the same one who set up the app. So, when the file picker tries to open the file explorer, it's like the app doesn't recognize the current user and crashes.

To make things work, I suggest installing the app normally, without using "Run as Administrator." Then, when you use the app, just run it regularly, without administrator rights. This way, it should work without any issues.

I hope this resolves your problem.

**Comment by tbrummel** [comment]
Wow - at least an hour wasted on this bug.  Yes, it's a bug.  It cannot be "by design" that an administrator cannot open a file.

This is 2023 and I really wanted to like WinUI3 but I think this is the last straw for me - I'm moving back to WPF. 

Calling MediaDevice.GetAudioRenderSelector() crashes if you run an .exe and you ARE NOT running "As adminstrator".  So, I have to run As Admin...

...and then calling picker.PickSingleFileAsync() will crash if you ARE running an .exe "As Administrator".  
Oh, yeah, endless loops -- I remember learning about these :-)

I do see that I could use a tool to generate code that will call the old Windows APIs from WinUI3, but I'm in complete agreement with others who have asked "Why would I do that?".  I can more easily drop back to WPF and get my app to actually run -- not only from VS on the dev machine, but on other PCs - without jumping through hoops!

I may post my experience to the WinUI3 official feedback channel but, in summary, I feel like right now WinUI3 in not a UI framework but a new sandbox framework.  It's not for serious desktop development as the design architecture seem to assume I'm ok with being shut down in the background or having resources taken away from my app -- but those are not an option for my app and there is apparently no way to allow the owner of the PC to make such decisions as opposed to the Framework.

**Comment by DarranRowe** [comment]
@tbrummel 
>This is 2023 and I really wanted to like WinUI3 but I think this is the last straw for me - I'm moving back to WPF.

Being slightly unhelpful but technically correct, this isn't a WinUI 3/Windows App SDK problem. Both MediaDevice and the pickers are part of the UWP APIs, which are provided by Windows itself. Mostly, people only use these for the first time when working with the WinUI 3 project templates and somehow see these as being part of the Windows App SDK.

As I have previously stated, the pickers have the exact same problems when being used from a Windows API application without the Windows App SDK being loaded or even referenced. What's more, I have no issues calling MediaDevice.GetAudioRenderSelector in an application running as a regular user.

![Screenshot 2023-09-24 110237](https://github.com/microsoft/WindowsAppSDK/assets/52577874/f82d80e8-67d6-4c77-bfda-ba64ae4d3fd6)

So yes, the file pickers are a problem. But I honestly don't see any issues with MediaDevice.

**Comment by tbrummel** [comment]
@DarranRowe

I realize that the media API I'm using is technically part of WinRT (which UWP extends).  The point is that developing a WinUI3 app is overly painful (and I do have significant experience developing Windows apps to compare with).  APIs that require developers to inject the main window handle remind me of the Win32 days - and I expected WinUI3 to be better than previous APIs.  It's honestly not - so far at least.  The moving target of the Windows "experience" they are trying to expose and, at the same time, support multi-platform development is certainly challenging and a contributing factor.  However, you'd have to admit that the evolution from WPF to WinRT to UWP to WinUI3 has been anything but smooth.  They all have their challenges, but WinUI3 leaves even experienced developers digging through forums trying to figure out simple stuff like how to prevent a crash using the file picker dialog.  Really?

Anyway, I don't want to hijack this thread with the GetAudioRenderSelector issue, but I'm interested in what you did.  Mine works until I build an x64 .exe and run that - you can't run it from VS.  I haven't yet built a simple test app to nail down the core issue, but I noticed running As Admin does avoid it.  I just feel like I'm wasting a lot more time investigating internal crashes than with previous frameworks.

**Comment by DarranRowe** [comment]
@tbrummel 

```
#include <Windows.h>
#include <winrt/Windows.Foundation.h>
#include <winrt/Windows.Foundation.Collections.h>
#include <winrt/Windows.Media.Devices.h>

#include <string>
#include <iostream>

int wmain()
{
	using namespace std;
	using namespace winrt::Windows::Media::Devices;

	wcout << wstring_view(MediaDevice::GetAudioRenderSelector()) << L"\n";
	return 0;
}
```

![Screenshot 2023-09-24 195326](https://github.com/microsoft/WindowsAppSDK/assets/52577874/637c0d6b-eba4-45b3-afea-f0b555ef0991)

I just call it and it works.

**Comment by tbrummel** [comment]
@DarranRowe

I wasn't trying to imply that that API never works.  Your example has left out all of the "UI" win WinUI3.  I have repros that work and repros that don't.  I've got one now that fails 1 in every 4 or 5 runs on one PC (never from VS) and fails 100% on another PC.  So, it looks like a threading/timing issue -- I'm guessing linked to the inner workings of the UI components of WinUI (as this is being executed by a WinUI control).  I'll post back if I get it narrowed down.  I do know that the crash is always on that API call (if it crashes).

I'm digging in to see if I can narrow down the root cause... but that is the point I was trying to make:  I'm wasting a lot of time with the WinUI3 iteration of the MS UI Framework trying to get simple things working.  I remember similar investigations with Avalon (WPF pre-release) so I'm not saying that WinUI won't get there eventually... but even without "issues", the code restrictions/sandboxing might just be too much for many desktop apps.  

I'm currently developing the app simultaneously in WInUI3, WPF w/ WinRT libs (media), and WPF native - so I can compare the experiences.

**Comment by tbrummel** [comment]
@DarranRowe 

Ok, so I started a new WinUI3 unpackaged app with the latest runtimes and then slowly added code/xaml to it until I could get it to crash.  Here it is: If I bind an Image to a Button in the XAML and do the call to MediaDevice.GetAudioRenderSelector() from the loaded event of that control, I'll get the app to crash about 20% of the time.  If I remove either the image(s) or the call to GetAudioRenderSelector, I cannot get it to crash.  

So, there is a timing/threading issue with either binding in general or possibly the async loading of images and other APIs.  I'm not convinced that this specific API call is key -- could be that anything I call at the "point" in the code might surface the issue -- but if I remove that call, I can't crash (and I don't have any other API calls in mind to try and really don't know if I want to invest any more time in this -- I'm moving back to WPF for now).

**Comment by bogdan-patraucean** [comment]
@AdamBraden , hello, some updates on this issue would be very appreciated.

**Comment by PavlikBender** [comment]
Any news? I have same problem

**Comment by PavlikBender** [comment]
Thanks to @castorix . I found workaround for me and made interpretation of it https://github.com/PavlikBender/Pickers

**Comment by Balkoth** [comment]
It's 2024 and Windows App SDK has no working file dialog. What an achievement.

**Comment by tbrummel** [comment]
Ah, seeing that new post reminded me that I never did follow up here.  

I rewrote my music server app in WPF and it runs perfectly on Win 10 & 11 and I can deploy it (copy the files to another PC) with ease!  I can also run it 24 hours a day -- and use memory!  It's amazing!  I think this WPF thing could catch on!

Or maybe I'm one of the few remaining developers who still write code they want to run on PCs and have no desire for a "UI framework" that throttles their memory usage and app performance.  If I were targeting a mobile device or needed to worry about batteries, I would welcome such features -- and I'd probably have little need for an open file dialog.

This could be a simple case of misinformation - I was under the impression that WinUI3 was for writing desktop apps.  Honestly, I wish it were, but I don't think a framework that arrests so much control from the developer is going to succeed for serious desktop apps.

**Comment by uixiu** [comment]
> It's 2024 and Windows App SDK has no working file dialog. What an achievement.

Also, the browse folder picker can't pick the same folder next. To enable the selection button, one should go trick the dialog buggy logic by going elsewhere and returning 😁. I went back to  Winforms (win32) dialogs in WinUI 3.0 ✌🏻

**Comment by hclazarin** [comment]
This was reported over 2 years ago, I can't believe there is still no acceptable solution or at least a reliable workaround. After 6 weeks working on a new major project facing issue after issue after issue with this SDK this is the last one, I give up. This framework is certainly a massive load of crap and will spank the architect who decided for it. I rather rollback to WPF and lose a few weeks than getting frustrated with nonsense every 3 days for the next few months.

**Comment by mominshaikhdevs** [comment]
@manodasanW @MikeHillberg It's really 2024. whats the latest update on this?

**Comment by smourier** [comment]
> This was reported over 2 years ago, I can't believe there is still no acceptable solution or at least a reliable workaround.

FWIW, you can use Winforms's OpenFileDialog (or WPF) easily from WinUI3 https://stackoverflow.com/a/78842081/403671

**Comment by mominshaikhdevs** [comment]
> What we could consider is creating a WinAppSDK wrapper around IFileDialog apis that provide an equivalent to the Picker apis. There are some opportunities to improve as well - StorageFiles are very heavyweight objects, and so we could consider returning something more optimized.

YES. This is the Right Solution that should have been provided by Windows App SDK _already_.
#
@AdamBraden @manodasanW @MikeHillberg 
It's really 2024. whats the latest update on this?

**Comment by Apollo199999999** [comment]
It's been 2 years and there's still no proper solution to this? Using WinForms OpenFileDialog as a workaround shouldn't be an acceptable solution.

**Comment by AlexanderBlackman** [comment]
This implementation works for me, but it's not on NuGet so you have to include it manually.
https://github.com/PavlikBender/Pickers



> It's been 2 years and there's still no proper solution to this? Using WinForms OpenFileDialog as a workaround shouldn't be an acceptable solution.

I feel your pain, I checked this thread again today when the WinUI 3 Gallery proudly announced it added a FilePicker sample, but no, it is still as broken as usual.

**Comment by sungaila** [comment]
The picker situation is aweful right now. For a packaged Desktop apps there are two options:

- Win32 pickers: do not work in AppContainer/PartialTrust (even with broadFileSystemAccess rights you cannot browse your computer)
- WinRT pickers: do not work when elevated/with admin rights (COMException with no context) and do not work in AppContainer/PartialTrust (WinUI 3 has no CoreWindow)

So the only option might be to build a custom picker that uses WinRT APIs, I guess? What a mess.

**Comment by Cacsjep** [comment]
when you think for an whole hour you are crazy and you doing something wrong and founding then such a huge github issue about a file dialog, what the hell

**Comment by tbrummel** [comment]
Over 2 years and still no fix?  This means that the project is ranked low in the dev team meetings and just never makes the cut for release.  This means that the framework is targeting mobile apps (that could maybe also run on a desktop).  I wish Microsoft would just make that clear.  There are a lot of decisions to make in choosing platforms and APIs for a project and letting people know that WinUI is not being developed as a desktop focused solution (but it might work for some scenarios) would be nice.   It's sad that developers are wasting time with debugging and investigation of issues that have been known for years -- and for what are clearly basic operations.  I mean if you're trying to inject a custom shader into the UI pipeline, you might expect to run into some issues -- but not when trying to open a file!  

I haven't revisited WinUI since my original post here and it has been on my "todo" list -- but I guess I can just remove it as there is, apparently, little progress in the direction I need for desktop/business apps/utilities.  This truly saddens me.  If/when I need to write a mobile photo viewer, I might check it out again...

**Comment by Rybasum** [comment]
@MartyIX Interesting... Do you know if this works in AppContainer? The test app included there has `runFullTrust` capability.

**Comment by mominshaikhdevs** [comment]
https://github.com/microsoft/WindowsAppSDK/discussions/4710#discussioncomment-11503257

**Comment by Rybasum** [comment]
@MartyIX https://github.com/microsoft/WindowsAppSDK/discussions/4710#discussioncomment-11503257
It looks like "the preview version of these APIs will not initially support app container scenarios". Thanks, Momin Shaikh!

**Comment by alski** [comment]
Just adding myself to the list wanting this

![Image](https://github.com/user-attachments/assets/04cf5615-b90d-4fa7-9c32-c0838baf20b7)

**Comment by ghost1372** [comment]
If anyone is interested, I have implemented FilePicker, FolderPicker, and SavePicker class that work in AOT, and Elevated apps.
[Docs are ready](https://ghost1372.github.io/DevWinUI/pickers/) and nuget package will be available soon.
https://github.com/ghost1372/DevWinUI/tree/main/dev/DevWinUI/Common/Picker

**Comment by ErcinDedeoglu** [comment]
I spent more than four hours getting rid of this error. Guys, you can't do this; there is no way you can leave it like this. This is a shame.
Fix it please.

**Comment by tbrummel** [comment]
I suspect that since most apps on mobile phones don't require the user to open a file, this isn't even on the radar.
Wait.  Open a file?  Isn't that a security violation?  😀
________________________________

**Comment by tbrummel** [comment]
That sounds like some good news.  Maybe I'll revisit the API at v2.  Thanks for the update
________________________________

**Comment by Balkoth** [comment]
> This isn't a Windows App SDK issue, this is a Windows issue that Microsoft has known about for a long time and just not fixed. This is talking about `Windows.Storage.Pickers.FileOpenPicker`.

I dont care which issue this is, it's happening here, was reported a long time ago and is the promoted solution when needing file dialogs. If they don't care for their tech to work the way desktop apps are supposed to work, then who else should care?

This is incredibly frustrating.

**Comment by DarranRowe** [comment]
> I dont care which issue this is, it's happening here, was reported a long time ago and is the promoted solution when needing file dialogs. If they don't care for their tech to work the way desktop apps are supposed to work, then who else should care?

The Windows App SDK team showed that **they do care**. They are going through the **best possible course of action that they can** if the WinRT team won't fix thing, by providing their own pickers.

**Comment by Balkoth** [comment]
You won't change my mind on this. It's four years late and their new pickers will have problems again from the start without AppContainer support. So again no unified solution when using WinAppSDK. People will use it within an AppContainer, find out it does not work and get frustrated again.

I am no longer buying this "We will fix it in the future" stuff.

**Comment by mominshaikhdevs** [comment]
regarding AppContainers support, still incomplete to this date.

Part Of: 
* https://github.com/microsoft/WindowsAppSDK/issues/219.

**Comment by GID0317** [comment]
is this fixed on Version 1.8.0 https://learn.microsoft.com/en-us/windows/apps/windows-app-sdk/stable-channel#storage-pickers? they said something like:

"The Microsoft.Windows.Storage.Pickers API in the Windows App SDK provides a modernized file and folder picker experience for desktop applications. This API is based on the existing Windows.Storage.Pickers API design, but with key improvements for desktop scenarios. The new Microsoft.Windows.Storage.Pickers API addresses two critical limitations of the UWP file and folder pickers on Apps developed with WinAppSDK/WinUI 3:

- Elevated Process Support: The existing Windows.Storage.Pickers APIs do not work when the application is running as an administrator. The new API enables file and folder selection in elevated mode.
- Simplified Usage in WinUI 3: Using the existing UWP pickers in WinUI 3 requires initializing a window handle for window association. The new pickers eliminate this requirement by accepting a WindowId directly in the constructor, making them easier to use."

**Comment by bogdan-patraucean** [comment]
@GID0317, I'm going to test and close the issue soon.

**Comment by alski** [comment]
Tested. Works perfectly while elevated (AKA Works on my machine)

<img width="1063" height="112" alt="Image" src="https://github.com/user-attachments/assets/3157c835-1738-4c6f-800c-3af8a3d36a66" />

```
var savePicker = new FileSavePicker(Root.MainWindow.AppWindow.Id)
    {
        SuggestedStartLocation = PickerLocationId.Desktop,
        SuggestedFileName = ...
    };
    savePicker.FileTypeChoices.Add("Json", new List<string>() { ".json" });

    //// Initialize the file picker with the window handle (HWND).
    //var hWnd = WinRT.Interop.WindowNative.GetWindowHandle(Root.MainWindow);
    //WinRT.Interop.InitializeWithWindow.Initialize(savePicker, hWnd);

    var file = await savePicker.PickSaveFileAsync();
    if (file != null)
    {       
        await ....Save(file.Path);
    }
```

**Comment by bogdan-patraucean** [comment]
Tested as well, took them only 3 years but at least we finally have it. I encourage you to open other issues if you see something missing or bugs, since the core feature seems to work. Thank you all for the engagement on this.

**Comment by Rybasum** [comment]
I tried the new file pickers with WinAppSDK 1.8 and 2.0-experimental4 and they don't work in AppContainer. That is, the dialog opens and immediately fails with Access Denied.

**Comment by DinahK-2SO** [comment]
> I tried the new file pickers with WinAppSDK 1.8 and 2.0-experimental4 and they don't work in AppContainer. That is, the dialog opens and immediately fails with Access Denied.

Hi @Rybasum , thanks for raising this point.

Right now, the new pickers are supported in containerized [AppContainer with win32-isolation](https://github.com/microsoft/win32-app-isolation). 

Please stay tuned for future updates as we continue to expand storage picker capabilities.

**Comment by tbrummel** [comment]
Big sigh.  How long has this new API been out in the wild?  How can an API that doesn't include file access even be released??
I mean isn't file read/write the second thing most programmers learned to write after "Hello World"?  It's just insane.

I've loved programming Windows since v3.11 and have been doing so ever since... but it seems Microsoft's vision is no longer to empower users and developers to use and program their own PCs.  It feels like it's now about connecting everyone to cloud services they may or may not want and making sure developers can't do anything that might cause "harm".  If an app had access to file APIs, it might be able to read information or even delete things! (I know that's what your program is supposed to do but...)

Each new API seems to remove, not add, functionality.  I just saw a post about how WinUI3 now has a wrap panel - yep, the same functionality we've had in WPF for over a decade.  It makes me sad.  How to we get back on track?



________________________________
